src/win32/fs.cpp

```diff
@@ -737,88 +701,111 @@ void WinFileSystemAccess::path2local(const string* path, string* local) const
     }
 }
 
+// convert UTF-8 to Windows Unicode
+void WinFileSystemAccess::path2local(const string* path, std::wstring* local) const
+{
+    // make space for the worst case
+    local->resize(path->size() + 2);
+
+    int len = MultiByteToWideChar(CP_UTF8, 0,
+        path->c_str(),
+        -1,
+        local->data(),
+        int(local->size()));
+    if (len)
+    {
+        // resize to actual result
+        local->resize(len - 1);
+    }
+    else
+    {
+        local->clear();
+    }
+}
+
 // convert Windows Unicode to UTF-8
 void WinFileSystemAccess::local2path(const string* local, string* path) const
 {
-    path->resize((local->size() + 1) * 4 / sizeof(wchar_t));
+    path->resize((local->size() + 1) * 4 / sizeof(wchar_t) + 1);
 
     path->resize(WideCharToMultiByte(CP_UTF8, 0, (wchar_t*)local->data(),
                                      int(local->size() / sizeof(wchar_t)),
                                      (char*)path->data(),
-                                     int(path->size() + 1),
+                                     int(path->size()),
                                      NULL, NULL));
     normalize(path);
 }
 
+void WinFileSystemAccess::local2path(const std::wstring* local, string* path) const
+{
+    path->resize((local->size() * sizeof(wchar_t) + 1) * 4 / sizeof(wchar_t) + 1);
+
+    path->resize(WideCharToMultiByte(CP_UTF8, 0, local->data(),
+        int(local->size()),
+        (char*)path->data(),
+        int(path->size()),
+        NULL, NULL));
+
+    normalize(path);
+}
+
 // write short name of the last path component to sname
 bool WinFileSystemAccess::getsname(LocalPath& namePath, LocalPath& snamePath) const
 {
 #ifdef WINDOWS_PHONE
     return false;
 #else
-    int r, rr;
-
-    string* name = namePath.editStringDirect();
-    string* sname = snamePath.editStringDirect();
+    const std::wstring& name = namePath.localpath;
+    std::wstring& sname = snamePath.localpath;
 
-    name->append("", 1);
+    DWORD r = DWORD(name.size());
+    sname.resize(r);
 
-    r = int(name->size() / sizeof(wchar_t) + 1);
+    DWORD rr = GetShortPathNameW(name.data(), sname.data(), r);
 
-    sname->resize(r * sizeof(wchar_t));
-    rr = GetShortPathNameW((LPCWSTR)name->data(), (LPWSTR)sname->data(), r);
-
-    sname->resize(rr * sizeof(wchar_t));
+    sname.resize(rr);
 
     if (rr >= r)
     {
-        rr = GetShortPathNameW((LPCWSTR)name->data(), (LPWSTR)sname->data(), rr);
-        sname->resize(rr * sizeof(wchar_t));
+        rr = GetShortPathNameW(name.data(), sname.data(), rr);
+        sname.resize(rr);
     }
 
-    name->resize(name->size() - 1);
-
     if (!rr)
     {
-        sname->clear();
+        DWORD e = GetLastError();
+        LOG_warn << "Unable to get short path name: " << namePath.localpath.c_str() << ". Error code: " << e;
```

Comment:
should we log wstring::c_str() `<< namePath.localpath.c_str()` . wouldn't we need a toPath here?

Reply:
good spotting - that would have just logged the address.   Updated